### PR TITLE
Finished first round of development

### PR DESCRIPTION
**Pull request by mostaphaRoudsari**
This PR addresses #15 

Now one can translate and run daylight-factor recipe using queenbee luigi. To successfully translate the recipe you need to provide the inputs.json for required inputs.

```
queenbee luigi translate path/to/daylight_factor.yaml path/to/project_folder -i inputs.json --workers 4
```

inputs.json
```JSON
{
    "radiance-parameters": "-I -ab 2 -h",
    "sensor-grid-count": "400",
    "input-grid": "model/grid/room.pts",
    "model": "model"
}
```

Put this model in project folder:  [model.zip](https://github.com/ladybug-tools/queenbee-luigi/files/4727126/model.zip)

Here is the daylight-factor recipe:

```yaml
metadata:
  name: daylight-factor
  version: latest
  keywords:
  - ladybug-tools
  - radiance
  maintainers:
  - name: mostapha
    email: mostapha@ladybug.tools
  home: null
  sources: null
  icon: https://www.ladybug.tools/assets/img/logo.png
  deprecated: null
  description: An awesome workflow to run daylight factor!
  license: null
dependencies:
- type: operator
  name: honeybee-radiance
  hash: 4e6b9354ade52b792988abef4426d744b587fb26eaabbc681ad27c75aa0c8048
  alias: null
  version: latest
  source: https://example.com/test-repo
flow:
- name: 6b0ac9f5b0c04f27d88656fef24d456a226fa34ea4524c24693f19445215ddb6/main
  inputs:
    parameters:
    - name: radiance-parameters
      default: -I -ab 2 -h
      description: The radiance parameters for ray tracing
      required: null
    - name: sensor-grid-count
      default: '100'
      description: The maximum number of grid points per parallel execution
      required: null
    artifacts:
    - name: input-grid
      description: A grid file
      default: null
      required: true
    - name: model
      description: A Honeybee model with radiance properties
      default: null
      required: true
  fail_fast: true
  tasks:
  - name: create-octree
    template: fc7f70871fc410a19274563857da4fc8c650ecac204c166b5359fafae09a4efc/create-octree
    arguments:
      parameters: []
      artifacts:
      - name: model
        from:
          type: inputs
          variable: model
        subpath: null
      - name: sky
        from:
          type: tasks
          name: generate-sky
          variable: sky
        subpath: null
    dependencies:
    - generate-sky
    loop: null
    sub_folder: null
    outputs:
      parameters: []
      artifacts:
      - name: scene-file
        path: output/octree/scene.oct
  - name: daylight-factor-simulation
    template: fc7f70871fc410a19274563857da4fc8c650ecac204c166b5359fafae09a4efc/ray-tracing
    arguments:
      parameters:
      - name: radiance-parameters
        from:
          type: inputs
          variable: radiance-parameters
        value: null
      artifacts:
      - name: grid
        from:
          type: tasks
          name: split-grid
          variable: output-grids-folder
        subpath: '{{item.path}}'
      - name: scene-file
        from:
          type: tasks
          name: create-octree
          variable: scene-file
        subpath: null
    dependencies:
    - split-grid
    - create-octree
    loop:
      from:
        type: tasks
        name: split-grid
        variable: grid-list
      value: null
      control:
        key: '{{item.name}}'
    sub_folder: null
    outputs:
      parameters: []
      artifacts:
      - name: result-file
        path: output/raw/{{item.name}}.res
  - name: generate-sky
    template: fc7f70871fc410a19274563857da4fc8c650ecac204c166b5359fafae09a4efc/10000-lux-sky
    arguments:
      parameters: []
      artifacts: []
    dependencies: null
    loop: null
    sub_folder: null
    outputs:
      parameters: []
      artifacts:
      - name: sky
        path: asset/sky/10000_lux.sky
  - name: post-process
    template: fc7f70871fc410a19274563857da4fc8c650ecac204c166b5359fafae09a4efc/post-process
    arguments:
      parameters: []
      artifacts:
      - name: raw-results-folder
        from:
          type: tasks
          name: daylight-factor-simulation
          variable: result-file
        subpath: null
    dependencies:
    - daylight-factor-simulation
    loop: null
    sub_folder: null
    outputs:
      parameters: []
      artifacts: []
  - name: split-grid
    template: fc7f70871fc410a19274563857da4fc8c650ecac204c166b5359fafae09a4efc/split-grid
    arguments:
      parameters:
      - name: grid-count
        from:
          type: inputs
          variable: sensor-grid-count
        value: null
      artifacts:
      - name: grid
        from:
          type: inputs
          variable: input-grid
        subpath: null
    dependencies: null
    loop: null
    sub_folder: null
    outputs:
      parameters:
      - name: grid-list
      artifacts:
      - name: output-grids-folder
        path: output/temp
  outputs:
    parameters: []
    artifacts:
    - name: raw-grids
      from:
        type: tasks
        name: daylight-factor-simulation
        variable: result-file
digest: 6b0ac9f5b0c04f27d88656fef24d456a226fa34ea4524c24693f19445215ddb6
templates:
- name: fc7f70871fc410a19274563857da4fc8c650ecac204c166b5359fafae09a4efc/10000-lux-sky
  description: Generates a 10000 lux sky
  inputs:
    parameters: []
    artifacts: []
  command: honeybee radiance sky illuminance 100000
  outputs:
    parameters: []
    artifacts:
    - name: sky
      description: The resulting sky object
      path: 100000_lux.sky
  config:
    docker:
      image: ladybugtools/honeybee-radiance:latest
      registry: null
      workdir: /opt/run-folder/
    local: null
- name: fc7f70871fc410a19274563857da4fc8c650ecac204c166b5359fafae09a4efc/create-octree
  description: Generate an octree
  inputs:
    parameters:
    - name: scene-files
      default: model/static/opaque/sample_case.mat model/static/opaque/sample_case.rad
        model/static/opaque/outdoor/context.mat model/static/opaque/outdoor/context.rad
        model/static/opaque/indoor/partition.mat model/static/opaque/indoor/partition.rad
        model/static/nonopaque/indoor/partition_glass.mat model/static/nonopaque/indoor/partition_glass.rad
      description: null
      required: false
    artifacts:
    - name: model
      description: null
      path: model
    - name: sky
      description: null
      path: sky.sky
  command: oconv -f sky.sky {{inputs.parameters.scene-files}} > scene.oct
  outputs:
    parameters: []
    artifacts:
    - name: scene-file
      description: null
      path: scene.oct
  config:
    docker:
      image: ladybugtools/honeybee-radiance:latest
      registry: null
      workdir: /opt/run-folder/
    local: null
- name: fc7f70871fc410a19274563857da4fc8c650ecac204c166b5359fafae09a4efc/post-process
  description: Post process your results and $$$
  inputs:
    parameters: []
    artifacts:
    - name: raw-results-folder
      description: null
      path: raw
  command: ls raw
  outputs:
    parameters: []
    artifacts: []
  config:
    docker:
      image: ladybugtools/honeybee-radiance:latest
      registry: null
      workdir: /opt/run-folder/
    local: null
- name: fc7f70871fc410a19274563857da4fc8c650ecac204c166b5359fafae09a4efc/ray-tracing
  description: Run ray tracing using some input data!
  inputs:
    parameters:
    - name: radiance-parameters
      default: -ab 2
      description: a string of radiance parameters
      required: false
    artifacts:
    - name: grid
      description: null
      path: grid.pts
    - name: scene-file
      description: null
      path: scene.oct
  command: rtrace -I -h {{inputs.parameters.radiance-parameters}} scene.oct < grid.pts > grid.res
  outputs:
    parameters: []
    artifacts:
    - name: result-file
      description: null
      path: grid.res
  config:
    docker:
      image: ladybugtools/honeybee-radiance:latest
      registry: null
      workdir: /opt/run-folder/
    local: null
- name: fc7f70871fc410a19274563857da4fc8c650ecac204c166b5359fafae09a4efc/split-grid
  description: null
  inputs:
    parameters:
    - name: grid-count
      default: null
      description: null
      required: true
    artifacts:
    - name: grid
      description: The input grid to split
      path: grid.pts
  command: honeybee radiance grid split grid.pts {{inputs.parameters.grid-count}}
    --folder output --log-file output/grids.txt
  outputs:
    parameters:
    - name: grid-list
      description: null
      path: output/grids.txt
    artifacts:
    - name: output-grids-folder
      description: null
      path: output
  config:
    docker:
      image: ladybugtools/honeybee-radiance:latest
      registry: null
      workdir: /opt/run-folder/
    local: null

```

You should see this graph at http://localhost:8082/ if you click on `LetItFly` task.

image

### CODE REVIEW

File: queenbee_luigi/recipe.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+        module_import = f'\nimport {subfolder}.main as workerbee'
+        if module_import not in self._imports:
+            self._imports.append(module_import)
+
+        contents = [self.imports, self._runner, self._footer(workers)]
+        filepath = f'run_{subfolder}.py'
+        with open(os.path.join(project_folder, filepath), 'w') as runner:
+            runner.write('\n\n\n'.join(contents).replace('\t', '    '))
+
+        if run:
```
Comment (AntoineDao) [inline queenbee_luigi/recipe.py:118]: I tried using the `--run` flag when generating a `run_daylight_factor.py` file in the current dir or in another dir (ie: `project_folder` = `.` or `./project_folder` and it failed for both.

The example below uses the following inputs, where `project_folder` is in the current dir:
```json
{
  "radiance-parameters": "-I -ab 2 -h",
  "sensor-grid-count": "400",
  "input-grid": "project_folder/asset/grid/room.pts",
  "model": "project_folder/model"
}
```
```console
> queenbee luigi translate test.yaml . -i inputs.json --workers 4 --run

Traceback (most recent call last):
  File "/home/kakistocrat/.pyenv/versions/3.6.6/bin/queenbee", line 8, in <module>
    sys.exit(main())
  File "/home/kakistocrat/.pyenv/versions/3.6.6/lib/python3.6/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/home/kakistocrat/.pyenv/versions/3.6.6/lib/python3.6/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/home/kakistocrat/.pyenv/versions/3.6.6/lib/python3.6/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/kakistocrat/.pyenv/versions/3.6.6/lib/python3.6/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/kakistocrat/.pyenv/versions/3.6.6/lib/python3.6/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/kakistocrat/.pyenv/versions/3.6.6/lib/python3.6/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/home/kakistocrat/.pyenv/versions/3.6.6/lib/python3.6/site-packages/click/decorators.py", line 21, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/home/kakistocrat/Projects/ladybug-tools/queenbee-luigi/queenbee_luigi/cli.py", line 58, in translate
    rep.write(project_folder=project_folder, workers=workers, values=values, run=run)
  File "/home/kakistocrat/Projects/ladybug-tools/queenbee-luigi/queenbee_luigi/recipe.py", line 121, in write
    subprocess.call(f'python {filepath}')
  File "/home/kakistocrat/.pyenv/versions/3.6.6/lib/python3.6/subprocess.py", line 267, in call
    with Popen(*popenargs, **kwargs) as p:
  File "/home/kakistocrat/.pyenv/versions/3.6.6/lib/python3.6/subprocess.py", line 709, in __init__
    restore_signals, start_new_session)
  File "/home/kakistocrat/.pyenv/versions/3.6.6/lib/python3.6/subprocess.py", line 1344, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
FileNotFoundError: [Errno 2] No such file or directory: 'python run_daylight_factor.py': 'python run_daylight_factor.py'
```

File: queenbee_luigi/dag.py [hunk 2]
```diff
@@ -17,114 +18,272 @@ def __init__(self, dag, templates):
...
-            cfg_file.write(f'project_folder: {project_folder}\n')
-            cfg_file.write(f'params_folder: {tmp_folder}\n')
-
-        tmp_folder = os.path.join(project_folder, tmp_folder)
-        if not os.path.isdir(tmp_folder):
-            os.makedirs(tmp_folder)
+    def write(self, folder='.', *, workers=1, values=None):
+        """Write DAG to folder.
+
+        A DAG will be written as a folder with subfolders for any possible DAG dependency
```
Comment (AntoineDao) [inline queenbee_luigi/dag.py:270]: Neat. Can't wait to try it out! :smile:

File: queenbee_luigi/recipe.py (hunk 1 above)
Comment (mostaphaRoudsari) [reply queenbee_luigi/recipe.py:118]: Hmm. I'll test this locally and see if I can create the issue. The way I put the path in was always relative to project folder but the way you have it should also work. Interesting that it didn't work.

**Comment by mostaphaRoudsari** [comment]
@AntoineDao, thank you for the feedback. Very good comments.

> Would be nice to point to a non-baked recipe folder and bake it within queenbee-luigi. This is how queenbee-argo works currently.

It is already supported. You can pass a folder instead of a file and it will create a baked-recipe from the folder.

> Would be nice to generate python files by default to a C:\Users\{username}\.queenbee\luigi\runs\tmp\<id> type folder so it doesn't pollute the user's folders when running it. 

Agreed! I tried doing that and hit the bug that I talked about before. Luigi fails to see the output files when they are finished. I will spend more time figuring this out once I have the recipes with DAG templates fully working.  https://github.com/ladybug-tools/queenbee-luigi/issues/23

> Would be nice to be able to generate the python files once per recipe and then inject the inputs.json into it as arguments when running it so each series of python files represents a Recipe rather than a workflow

That is how it was set up but I decided against it not to make them reusable. Mainly for non-technical reasons. 😁. We can chat about this.

> I'm not sure what the story is between running and visualising recipes at the moment. It seems we have to run luigid in the background to be able to view it. Is there a way to integrate these command into queenbee luigi cli per chance?

I'm trying to do this here when you run the translate command but it looks like that it didn't work on your machine. The problem is that running luigi daemon in background is not a straight forward task in Windows. We can manage to do it by some tricks of course. I will work on it. For now, one has to open a CMD and run `luigid`.

**Pull request merged**